### PR TITLE
Make remove backup data optional

### LINKED ISSUE #2888: backups block namespace deletion
## Bug Report

**What version of Kubernetes are you using?**

**What version of TiDB Operator are you using?**
v1.1.0

**What did you do?**
I deleted the namespace the `TidbCluster` was in.

**What did you expect to see?**

The namespace eventually gets deleted

**What did you see instead?**

The namespace got stuck in the `Terminating` state. Here's what I think happened:

There were `Backups` for the cluster. To delete them, the operator first runs a `Job` to delete the data from S3. However, since the namespace is `Terminating`, Kubernetes refuses to create the `Job`. The operator will keep retrying, and the finalizer on the `Backup` prevents it from being deleted. So the namespace just ends up stuck.

The only workaround seems to be to delete the finalizers.

[operator-logs.txt](https://github.com/pingcap/tidb-operator/files/4891954/operator-logs.txt)

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
Fix https://github.com/pingcap/tidb-operator/issues/2286 Fix #2888 

### What is changed and how does it work?
Add `cleanData` argument in backup spec(default: false). When it's enabled, the dumped data will be deleted after backup process is deleted.

### Check List 

Tests 

 - Unit test
 - E2E test
 - Manual test (add detailed scripts or steps below)
   * Test backup to S3 when `cleanData` is false. After deletion dumped data won't be deleted.
   * Test backup to S3 when `cleanData` is true. After deletion dumped data will be deleted.

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
ACTION REQUIRED: Add a field `cleanData` in `BackupSpec` to denote whether to clean backup data before the object is deleted from the cluster (defaults to false). If you want to clean backup data as before, please set `spec.cleanData` in `Backup` yaml or `spec.backupTemplate.cleanData` in `BackupSchedule` yaml to true.
```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 1]
```diff
@@ -1138,6 +1138,8 @@ type BackupSpec struct {
 	UseKMS bool `json:"useKMS,omitempty"`
 	// Specify service account of backup
 	ServiceAccount string `json:"serviceAccount,omitempty"`
+	// CleanData denotes whether to clean backuped data after backup job is deleted
```
Comment (cofyc) [inline pkg/apis/pingcap/v1alpha1/types.go:1141]:
```suggestion
	// CleanData denotes whether to clean backup data before the object is deleted from the cluster
```

the finalizer implements `pre-delete` hooks.

**Comment by lichunzhu** [comment]
@DanielZhangQD @cofyc PTAL

File: pkg/controller/backup/backup_control.go [hunk 2]
```diff
@@ -53,12 +53,14 @@ type defaultBackupControl struct {
 // UpdateBackup executes the core logic loop for a Backup.
 func (bc *defaultBackupControl) UpdateBackup(backup *v1alpha1.Backup) error {
 	backup.SetGroupVersionKind(controller.BackupControllerKind)
-	if err := bc.addProtectionFinalizer(backup); err != nil {
-		return err
-	}
+	if backup.Spec.CleanData {
```
Comment (cofyc) [inline pkg/controller/backup/backup_control.go:56]: this check seems not necessary and IMO we should always try to add/remove finalizer, for example, users may change this field from `true` to `false`

**Comment by cofyc** [comment]
a simplified version of release note, what do you think?

```
Add a field `spec.cleanData` in `Backup` to denote whether to clean backup data before the object is deleted from the cluster  (defaults to false)
```

**Comment by lichunzhu** [comment]
> a simplified version of release note, what do you think?
> 
> ```
> Add a field `spec.cleanData` in `Backup` to denote whether to clean backup data before the object is deleted from the cluster  (defaults to false)
> ```

Revised. It's much better.

**Comment by cofyc** [comment]
oh, this is an incompatible change. The old Backup will always try to clean the data.

I think we need an `ACTION REQUIRED` release note and telling people to set the `cleanData` field to `true` if they want to keep the old behavior.

**Comment by cofyc** [comment]
cc @DanielZhangQD do we need to cherry-pick this into 1.1?

**Comment by lichunzhu** [comment]
> oh, this is an incompatible change. The old Backup will always try to clean the data.
> 
> I think we need an `ACTION REQUIRED` release note and telling people to set the `cleanData` field to `true` if they want to keep the old behavior.

Yes, you're right.

**Comment by DanielZhangQD** [comment]
> cc @DanielZhangQD do we need to cherry-pick this into 1.1?

Yes, I have added the label.

**Comment by lichunzhu** [comment]
> oh, this is an incompatible change. The old Backup will always try to clean the data.
> 
> I think we need an `ACTION REQUIRED` release note and telling people to set the `cleanData` field to `true` if they want to keep the old behavior.

Added in release note.

**Comment by cofyc** [comment]
@lichunzhu please also set `cleanData` to `true` in `GetBackupCRDWithS3` (`tests/pkg/fixture/fixture.g`). We should clean backup data in our testing.

File: pkg/controller/backup/backup_control.go [hunk 3]
```diff
@@ -86,7 +86,7 @@ func (bc *defaultBackupControl) removeProtectionFinalizer(backup *v1alpha1.Backu
 	ns := backup.GetNamespace()
 	name := backup.GetName()
 
-	if isDeletionCandidate(backup) && v1alpha1.IsBackupClean(backup) {
+	if !backup.Spec.CleanData || (isDeletionCandidate(backup) && v1alpha1.IsBackupClean(backup)) {
```
Comment (DanielZhangQD) [inline pkg/controller/backup/backup_control.go:89]:
```suggestion
	if backup.Spec.CleanData && isDeletionCandidate(backup) && v1alpha1.IsBackupClean(backup) {
```

Comment (DanielZhangQD) [reply pkg/controller/backup/backup_control.go:89]: Is my suggestion correct?

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 4]
```diff
@@ -1138,6 +1138,8 @@ type BackupSpec struct {
 	UseKMS bool `json:"useKMS,omitempty"`
 	// Specify service account of backup
 	ServiceAccount string `json:"serviceAccount,omitempty"`
+	// CleanData denotes whether to clean backup data before the object is deleted from the cluster
```
Comment (DanielZhangQD) [inline pkg/apis/pingcap/v1alpha1/types.go:1141]: Please help update the examples in `manifests/backup/`.

Comment (lichunzhu) [reply pkg/apis/pingcap/v1alpha1/types.go:1141]: addressed in https://github.com/pingcap/tidb-operator/pull/2892/commits/ee29d324b6fc2e46c4587c2bb4ecd9a83947ad4d

File: pkg/controller/backup/backup_control.go (hunk 3 above)
Comment (lichunzhu) [reply pkg/controller/backup/backup_control.go]: addressed in https://github.com/pingcap/tidb-operator/pull/2892/commits/156e00d9240831138b91865b96c2dcea9a600d10

File: manifests/backup/backup-aws-s3-br.yaml [hunk 5]
```diff
@@ -10,6 +10,7 @@ spec:
   # backupType: full
   # useKMS: false
   # serviceAccount: myServiceAccount
+  # cleanData: true
```
Comment (DanielZhangQD) [inline manifests/backup/backup-aws-s3-br.yaml:13]: No need for BR yamls.

Comment (DanielZhangQD) [reply manifests/backup/backup-aws-s3-br.yaml:13]: sorry, it also needs to change for BR.

**Comment by DanielZhangQD** [comment]
@lichunzhu The release notes also have to mention `BackupSchedule`CR.

**Comment by lichunzhu** [comment]
> @lichunzhu The release notes also have to mention `BackupSchedule`C

> @lichunzhu The release notes also have to mention `BackupSchedule`CR.

Release note has been revised. Conflicts have also been resolved.

**Pull request merged**